jxl/src/render/stages/upsample.rs

```diff
@@ -57,6 +80,410 @@ impl<const N: usize, const SHIFT: u8> std::fmt::Display for Upsample<N, SHIFT> {
     }
 }
 
+/// State for upsampling stage containing reusable temp buffers
+struct UpsampleState {
+    col_min: Vec<f32>,
+    col_max: Vec<f32>,
+    mins: Vec<f32>,
+    maxs: Vec<f32>,
+}
+
+impl UpsampleState {
+    fn new() -> Self {
+        Self {
+            col_min: Vec::new(),
+            col_max: Vec::new(),
+            mins: Vec::new(),
+            maxs: Vec::new(),
+        }
+    }
+
+    fn ensure_capacity(&mut self, xsize: usize) {
+        // Generous padding for SIMD operations:
+        // Step 1: needs xsize + 4 (col_fill_len) elements
+        // Step 2: loads at offset + 4, needs SIMD_LEN more elements from there
+        // For AVX-512 (SIMD_LEN=16), we need: xsize + 4 + 16 = xsize + 20
+        // Add extra padding to be safe with future SIMD widths
+        let needed = xsize + 24;
+        if self.col_min.len() < needed {
+            self.col_min.resize(needed, 0.0);
+            self.col_max.resize(needed, 0.0);
+            self.mins.resize(needed, 0.0);
+            self.maxs.resize(needed, 0.0);
+        }
+    }
+}
+
+// Shared min/max computation for all upsampling variants
+simd_function!(
+    compute_minmax_simd_dispatch,
+    d: D,
+    fn compute_minmax_simd(
```

Comment:
I don't think this needs its own dispatch - you can make a single function that gets dispatched, which then calls a function that is generic on D (or two -- one for minmax and one for the actual upsampling).

This makes the body in the macro small, and is somewhat more efficient than doing two dynamic dispatches.